controlplane/kubeadm/internal/etcd/etcd.go

```diff
@@ -0,0 +1,233 @@
+	MemberID uint64
+
+	// Type is the type of alarm which has been raised.
+	Type AlarmType
+}
+
+type AlarmType int32
+
+const (
+	// AlarmOK denotes that the cluster member is OK.
+	AlarmOk AlarmType = iota
+
+	// AlarmNoSpace denotes that the cluster member has run out of disk space.
+	AlarmNoSpace
+
+	// AlarmCorrupt denotes that the cluster member has corrupted data.
+	AlarmCorrupt
+)
+
+// Adapted from kubeadm
+
+// Member struct defines an etcd member; it is used to avoid spreading
+// github.com/coreos/etcd dependencies.
+type Member struct {
+	// ID is the ID of this cluster member
+	ID uint64
+
+	// Name is the human-readable name of the member. If the member is not started, the name will be an empty string.
+	Name string
+
+	// PeerURLs is the list of URLs the member exposes to the cluster for communication.
+	PeerURLs []string
+
+	// ClientURLs is the list of URLs the member exposes to clients for communication. If the member is not started, clientURLs will be empty.
+	ClientURLs []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// Alarms is the list of alarms for a member.
+	Alarms []AlarmType
+}
+
+// pbMemberToMember converts the protobuf representation of a cluster member to a Member struct.
+func pbMemberToMember(m *etcdserverpb.Member) *Member {
+	return &Member{
+		ID:         m.GetID(),
+		Name:       m.GetName(),
+		PeerURLs:   m.GetPeerURLs(),
+		ClientURLs: m.GetClientURLs(),
+		IsLearner:  m.GetIsLearner(),
+		Alarms:     []AlarmType{},
+	}
+}
+
+// NewEtcdClient creates a new etcd client with a custom dialer and is configuration with optional functions.
+func NewEtcdClient(endpoint string, dialer GRPCDial, tlsConfig *tls.Config) (*clientv3.Client, error) {
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{endpoint},
+		DialTimeout: etcdTimeout,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // block until the underlying connection is up
+			grpc.WithContextDialer(dialer),
+		},
+		TLS: tlsConfig,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create etcd client")
+	}
+	etcdClient.Endpoints()
+	return etcdClient, nil
+}
+
+// NewClientWithEtcd configures our response formatter (Client) with an etcd client and endpoint.
+func NewClientWithEtcd(etcdClient etcd) (*Client, error) {
+	if len(etcdClient.Endpoints()) == 0 {
+		return nil, errors.New("etcd client was not configured with any endpoints")
+	}
+	return &Client{
+		Endpoint:   etcdClient.Endpoints()[0],
+		EtcdClient: etcdClient,
+	}, nil
+}
+
+// Close closes the etcd client.
+func (c *Client) Close() error {
+	return c.EtcdClient.Close()
+}
+
+// Member retrieves the Member information for a given peer.
+func (c *Client) Member(ctx context.Context, peerURL string) (*Member, error) {
+	members, err := c.Members(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range members {
+		if m.PeerURLs[0] == peerURL {
```

Comment:
I think we should use the member name to identify the member in the list... But if we use peerURL, then we should iterate over all peerURLs here.

Reply:
I'll update this to use name if you verify you need this functionality at all.

I think we should try to keep the interface as simple as possible without trying to make it overly performant before we decide this is actually a performance problem. The signature `MemberForNames(members []*Members, name string) *Member` indicates that you'd be keeping a reference to a list of members outside this API, something like this:

```
members := client.Members()
// some more work happens here
client.MemberForName(members, name)
```

But this has some issues due to the members list potentially becoming stale.

Let's make many API calls and if this becomes a bottleneck we can implement some kind of caching layer or redesign the API. We have a lot of flexibility here since it's in the `internal/` package.

Reply:
> I think we should use the member name to identify the member in the list

depending on the state of the joining members, they can end up being without a name.
URL or ID are unique at all times, so IMO any function used to "get" a member from a cached list should not use name.

Reply:
The intent was not to do any caching here, and any time you call into this client, it's explicitly going to go out and make some requests. How we choose to cache that is the consumer's concern.

Reply:
> depending on the state of the joining members, they can end up being without a name.

You make a good point, but I want to clarify it a little: When a member is added, it has no name. When the member _joins_, it gets a name. So a member without a name is unstarted.

> so IMO any function used to "get" a member from a cached list should not use name.

I didn't consider this function taking a "cached list" as input, because the user owns the list, not the etcd client.

The control plane controller will need to find a Member by its name, because the name is the one piece of information associated with a Machine (via its NodeRef). The member ID is not associated with a Machine in any way.

Reply:
> I'll update this to use name if you verify you need this functionality at all.

Good question :slightly_smiling_face:  In fact, I do _not_ need `Member` in this client for scale up or down.